Temporal Nexus: adding Nexus terms in Glossary (#3046)
temporalio/documentation: docs/glossary.md

_Tags: [term](/tags/term)_

> #### [Caller](/nexus#caller)

Author: removing

> The initiator or caller of a Nexus Operation, also known as the service consumer calling a service API. A Caller is also known as a source or consumer. 

Reviewer: as above, would skip this for the Temporal glossary

```suggestion
```

_Tags: [term](/tags/term), [explanation](/tags/explanation)_

> #### [Handler](/nexus#handler)

Reviewer 2: Hmmmmm, are we generally calling this a "handler" and not a "Nexus handler"? We use "handles" a lot in Core SDKs.

> #### [Handler](/nexus#handler)

Reviewer: would skip this one and keep the Nexus Operation Handler definition lower in the glossary.

```suggestion
```

Author: removing

> The Handler of a Nexus Operation, running in a Worker that is polling a Task Queue the target/destination Namespace, and associated underlying Temporal primitives like a Workflow. The Handler is also known as a callee, target, destination, or provider. 

Reviewer: as above, would skip this for the Temporal glossary

```suggestion
```

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Operation Events](/nexus#nexus-operation-events)

History events that surface in the Caller Workflow to indicate the state of an Operation including `Nexus Operation Scheduled`, `Nexus Operation Started`, `Nexus Operation Completed`.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Operation Handler](/nexus#nexus-operation-handler)

The Nexus handler code in a Temporal Worker typically created using Temporal SDK builder functions that make it easy to abstract Temporal primitives and expose a clean service contract for others to use.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Service](/nexus#nexus-service)

A Nexus Service is a named collection of arbitrary-duration Nexus Operations that provide an API contract suitable for sharing across team boundaries.
Nexus Services are registered with a Temporal Worker that is listening on the target Namespace and Task Queue for an Endpoint.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

> 

Reviewer: ```suggestion
#### [Nexus RPC](https://github.com/nexus-rpc/api/blob/main/SPEC.md)
Nexus RPC is a new protocol designed with durable execution in mind. It supports arbitrary-duration Operations that extend beyond a traditional RPC — a key underpinning to connect durable executions within and across namespaces, clusters, regions, and clouds.
```

> 

Reviewer: need to check the link, but that page has details on the Nexus Machinery

```suggestion
#### [Nexus Machinery](/cloud/nexus/using-nexus)
In Temporal Cloud, the Nexus Machinery routes Nexus requests across Namespace boundaries via a secure mTLS Envoy mesh, where they are sync matched to a Worker in the handler Namespace, which processes the Nexus request.
```

#### [Namespace](/namespaces)

A Namespace is a unit of isolation within the Temporal Platform.

_Tags: [term](/tags/term), [explanation](/tags/explanation)_

#### [Nexus API Definitions](/nexus#nexus-api-defintions)

A common code package, schema, or documentation that a Caller can use to obtain Service and Operation names as associated input/output types a Service will accept for a given Operation.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus API Registry](/nexus#nexus-api-registry)

An account-scoped list of Endpoints and Services available for use within a Temporal Cloud account, including an Endpoint description of the Services and Operations which the Endpoint supports. 

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

> #### [Nexus Client](/nexus#nexus-client)

Reviewer: Go SDK specific, would remove

```suggestion
```

Author: removing

Used by the Caller to connect to a Nexus Endpoint and Nexus Service.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Endpoint](/nexus#nexus-endpoint)

> A Nexus Endpoint is like an API proxy that can serve one or more Nexus Services from a target such as a Worker listening on a target Temporal Namespace and target Task Queue. 

Reviewer 2: Curious (possibly stupid question) -- what makes it "like" an API proxy rather than just "an API proxy"?

Reviewer: Would be fine removing "like" if we're comfortable with "API proxy" in this context translating to a  Temporal Nexus API proxy. Calling it an API proxy does make the concept immediately understandable to most everyone we have talked with.

Why "like"? Most traditional API proxies are typical HTTP request/response, whereas a Temporal Nexus Endpoint routes via Temporal's queue-based worker architecture via a namespace + task queue (and some Nexus RPC internally across some hops), so it didn't feel correct to position it as a traditional API proxy, but also it does perform the functions of an API proxy, perhaps just with a Temporal transport for part of the journey from caller to handler.

Also note that a Temporal Nexus Endpoint is more than just the low-level [Nexus endpoint](https://github.com/nexus-rpc/api/blob/30ebcf5ded186ccfc70759f864ce8a43d6c73859/SPEC.md?plain=1#L17), which does not route to Temporal namespaces, and Nexus RPC is agnostic to Temporal entirely -- much like a [Google Cloud Endpoint](https://cloud.google.com/endpoints) is a gRPC endpoint, but is also doing their own API proxy/gateway things, and gRPC is agnostic to Google Cloud Endpoints.

Author: Thanks for the context. Going to removing the word "like" for now unless there's an alternative.

This allows service providers to present a clean API contract and hide the underlying implementation, which may consist of many internal Workflows. 
Multiple Nexus Endpoints can target the same Namespace, and over time a Nexus Endpoint will be able to span multiple Namespaces with service routing rules.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Operation](/nexus#nexus-operation)

> An arbitrary-length service call that may be synchronous or asynchronous, short-lived, or long-lived, used to invoke operations across Namespace, Temporal Service, and cell boundaries and into arbitrary external systems.

Reviewer: ```suggestion
An arbitrary-duration operation that may be synchronous or asynchronous, short-lived, or long-lived, and used to connect durable executions within and across namespaces, clusters, regions, and clouds. Unlike a traditional RPC, an asynchronous Nexus Operation has an operation identity that can be used to re-attach to a long-lived Nexus Operation, for example, one backed by a Temporal Workflow. Nexus Operations support a uniform interface to get the status of an operation or its result, receive a completion callback, or cancel the operation – all of which are fully integrated into the Temporal platform.
```

Reviewer 3: I like what you've added here. I would change "re-attach to a long-lived Nexus Operation" to "re-attach to it".

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Operation Events](/nexus#nexus-operation-events)

> History events that surface in the Caller Workflow to indicate the state of an Operation including `Op Scheduled`, `Op Started`, `Op Completed`.

Reviewer: ```suggestion
History events that surface in the Caller Workflow to indicate the state of an Operation including `Nexus Operation Scheduled`, `Nexus Operation Started`, `Nexus Operation Completed`.
```

Reviewer 3: We don't have a section for activity events, I don't think we need one for Nexus.

Author: It doesn't hurt to have more definitions & it helps out with Ask AI purposes.

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

> #### [Nexus Operation Handle](/nexus#nexus-operation-handle)

Reviewer: Would omit this here, too low level and the examples will be SDK specific
```suggestion
```

> Returned to the Workflow that started an operation to interact with the operation, for example `handle.WaitStarted()`, `handle.GetResult()`.

Reviewer: ```suggestion
```

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Operation Handler](/nexus#nexus-operation-handler)

> The handler code in a Handler Worker, used to process service calls for a given Operation. Operation Handlers often run in the same Worker as the Workflows they abstract, to provide a service interface that can be used across Namespace boundaries.

Reviewer: ```suggestion
The Nexus handler code in a Temporal Worker typically created using Temporal SDK builder functions that make it easy to abstract Temporal primitives and expose a clean service contract for others to use.
```

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

> #### [Nexus Operation Reference](/nexus#nexus-operation-reference)

Reviewer: outdated

```suggestion
```

Reviewer 3: It's not outdated, it's just a Go specific concept.

> Key information about an Operation including the name and input/output types sufficient for invoking the Operation from a Caller Workflow, or implementing a Handler for the Operation in a Handler Worker.

Reviewer: as above

```suggestion
```

_Tags: [term](/tags/term), [nexus](/tags/nexus), [explanation](/tags/explanation)_

#### [Nexus Service](/nexus#nexus-service)

> Nexus Services are a Temporal primitive that supports arbitrary-duration Nexus Operations compatible with durable execution systems. 

Reviewer: ```suggestion
A Nexus Service is a named collection of arbitrary-duration Nexus Operations that provide an API contract suitable for sharing across team boundaries.
```

> Unlike a traditional RPC, an arbitrary-length Nexus Operation has an identity (id) that lasts beyond the process lifetimes of the Caller and Handler, and can be used to re-attach to a long-lived Nexus Operation, for example, one backed by a Temporal Workflow.

Reviewer: ```suggestion
Nexus Services are registered with a Temporal Worker that is listening on the target namespace and task queue for an Endpoint.
```
